features/storage/kvstore/DirectAccessDevicekey.cpp

```diff
@@ -0,0 +1,189 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------- Includes -----------------------------------------------------------
+
+#include "DirectAccessDevicekey.h"
+#include <string.h>
+#include <stdio.h>
+#include "mbed_error.h"
+#include "MbedCRC.h"
+#include "mbed_trace.h"
+#define TRACE_GROUP "DADK"
+
+using namespace mbed;
+
+// --------------------------------------------------------- Definitions ----------------------------------------------------------
+#define TDBSTORE_NUMBER_OF_AREAS 2
+#define MAX_DEVICEKEY_DATA_SIZE 64
+#define RESERVED_AREA_SIZE (MAX_DEVICEKEY_DATA_SIZE+8) /* DeviceKey Max Data size + 4bytes address + 4bytes deviceKey actual size */
+
+typedef struct {
+    uint32_t address;
+    size_t   size;
+} tdbstore_area_data_t;
+
+typedef struct {
+    uint16_t trailer_size;
+    uint16_t data_size;
+    uint32_t crc;
+} reserved_trailer_t;
+
+// -------------------------------------------------- Local Functions Declaration ----------------------------------------------------
+static int calc_area_params(BlockDevice *bd, uint32_t tdb_start_offset, uint32_t tdb_end_offset,
+                            tdbstore_area_data_t *area_params);
+static int reserved_data_get(BlockDevice *bd, tdbstore_area_data_t *area_params, void *reserved_data_buf,
+                             size_t reserved_data_buf_size, size_t *actual_data_size_ptr);
+static uint32_t calc_crc(uint32_t init_crc, uint32_t data_size, const void *data_buf);
+
+// -------------------------------------------------- API Functions Implementation ----------------------------------------------------
+int direct_access_to_devicekey(BlockDevice *bd, uint32_t tdb_start_offset, uint32_t tdb_end_offset, void *data_buf,
+                               size_t data_buf_size, size_t *actual_data_size_ptr)
+{
+    int status = MBED_ERROR_INVALID_ARGUMENT;
+    uint8_t active_area = 0;
+    tdbstore_area_data_t area_params[TDBSTORE_NUMBER_OF_AREAS];
+    memset(area_params, 0, sizeof(area_params));
+
+    if (NULL == data_buf) {
+        tr_error("Invalid Data Buf Argument");
+        goto exit_point;
+    }
+
+    status = calc_area_params(bd, tdb_start_offset, tdb_end_offset, area_params);
+    if (status != MBED_SUCCESS) {
+        tr_error("Couldn't calulate Area Params - err: %d", status);
+        goto exit_point;
+    }
+
+    /* DeviceKey data can be found either in first or second Flash Area */
+    /* Loop through Areas to find valid DeviceKey data */
+    for (active_area = 0; active_area < TDBSTORE_NUMBER_OF_AREAS; active_area++) {
```

Comment:
Would change the name to area (instead of active_area, which is a bit confusing).